### ISSUE TITLE
strictNullChecks false positives for non-null object properties

### ISSUE THREAD

**Issue by runny-yolk** [microsoft/TypeScript#60847]
### 🔎 Search Terms

possibly null undefined object property

### 🕗 Version & Regression Information

Originally detected with 5.7.2, but also confirmed in v5.8.0-dev.20241223 via the playground.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=5.8.0-dev.20241223#code/DYUwLgBAHgXBB2BXAtgIxAJwgHwY4wEAvBAKwDcAUKJAJ7HQQDUZV14EA9qgFZwDesPGkw48BAL4NBcUhKo0IALwbceAOijNWEAPS6ImDJwxwA5Gs1mIASwDOEAA6c7dm6mD0zSAmfUAKAEYADgAGABYAdgBKIA

### 💻 Code

```ts
let x: number | null = 5;
let y = x + 5; // no problem

let obj: {x: number | null} = {x: 5};
let z = obj.x + 5; // error: 'obj.x' is possibly 'null'. (18047)
```

### 🙁 Actual behavior

When nullable object properties have been assigned a non-null value, TS still considers them "possibly null", even though it does not for equivalently typed variables.

### 🙂 Expected behavior

No error.

**Comment by jcalz** [comment]
unions narrow, objects with union-typed properties do not; you'd need #42384 for this to work as desired. If you made `obj` of type `{x: number} | {x: null}` it would also work today (since that's a union of objects).
